cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operations/OperationRunsStore.java

```diff
@@ -0,0 +1,360 @@
+  public void updateOperationStatus(OperationRunId runId, OperationRunStatus status,
+      @Nullable byte[] sourceId) throws OperationRunNotFoundException, IOException {
+    OperationRunDetail<?> currentDetail = getRunDetail(runId);
+    OperationRun currentRun = currentDetail.getRun();
+    OperationRun updatedRun = OperationRun.builder(currentRun)
+        .setStatus(status).build();
+    OperationRunDetail<?> updatedDetail = OperationRunDetail.builder(currentDetail)
+        .setRun(updatedRun).setSourceId(sourceId).build();
+
+    Collection<Field<?>> fields = getCommonUpdateFields(runId);
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.DETAILS_FIELD,
+            GSON.toJson(updatedDetail)));
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.STATUS_FIELD,
+            GSON.toJson(status)));
+
+    StructuredTable operationRunsTable = getOperationRunsTable(context);
+    operationRunsTable.update(fields);
+  }
+
+  /**
+   * Mark an operation run as failed.
+   *
+   * @param runId {@link OperationRunId} for the run
+   * @param error error related to the failure
+   * @param sourceId the message id which is responsible for the update
+   * @throws OperationRunNotFoundException run with id does not exist in namespace
+   */
+  public void failOperationRun(OperationRunId runId, OperationError error,
+      @Nullable byte[] sourceId) throws OperationRunNotFoundException, IOException {
+    OperationRunDetail<?> currentDetail = getRunDetail(runId);
+    OperationRun currentRun = currentDetail.getRun();
+    OperationRun updatedRun = OperationRun.builder(currentRun)
+        .setStatus(OperationRunStatus.FAILED).setError(error).build();
+    OperationRunDetail<?> updatedDetail = OperationRunDetail.builder(currentDetail)
+        .setRun(updatedRun).setSourceId(sourceId).build();
+
+    Collection<Field<?>> fields = getCommonUpdateFields(runId);
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.DETAILS_FIELD,
+            GSON.toJson(updatedDetail)));
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.STATUS_FIELD,
+            GSON.toJson(OperationRunStatus.FAILED)));
+
+    StructuredTable operationRunsTable = getOperationRunsTable(context);
+    operationRunsTable.update(fields);
+  }
+
+  /**
+   * Get an operation run by id.
+   *
+   * @param runId {@link OperationRunId} for the run
+   * @throws OperationRunNotFoundException run with id does not exist in namespace
+   */
+  public OperationRunDetail<?> getOperation(OperationRunId runId)
+      throws OperationRunNotFoundException, IOException {
+    Optional<StructuredRow> row = getOperationRunInternal(runId);
+    if (!row.isPresent()) {
+      throw new OperationRunNotFoundException(runId.getNamespace(), runId.getRun());
+    }
+    return GSON.fromJson(
+        row.get().getString(StoreDefinition.OperationRunsStore.DETAILS_FIELD),
+        OperationRunDetail.class
```

Comment:
I'm still not sure how this will work without a concrete class for the generic. Maybe it will be more clear when there is code around the caller using the result, but I get the feeling it will need to revisited.